metric_learn/scml.py

```diff
@@ -0,0 +1,647 @@
+                         ' match the dimensionality of the given inputs `X` '
+                         '({}).'.format(basis.shape[1], n_features))
+    elif self.basis not in self._authorized_basis:
+      raise ValueError(
+          "`basis` must be one of the options '{}' "
+          "or an array of shape (n_basis, n_features)."
+          .format("', '".join(self._authorized_basis)))
+    if self.basis == 'triplet_diffs':
+      basis, n_basis = self._generate_bases_dist_diff(triplets, X)
+
+    return basis, n_basis
+
+  def _generate_bases_dist_diff(self, triplets, X):
+    """ Constructs the basis set from the differences of positive and negative
+    pairs from the triplets constraints.
+
+    The basis set is constructed iteratively by taking n_features positive and
+    n_features negative pairs differences, then adding and substracting
+    respectively all the outerproducts, and finally adding the eigenvectors
+    of this matrix with positive eigenvalue.
+    """
+    n_features = X.shape[1]
+    n_triplets = triplets.shape[0]
+
+    if self.n_basis is None:
+      # TODO: Get a good default n_basis directive
+      n_basis = int(n_triplets/10)
+      warnings.warn('As no value for `n_basis` was selected, the number of '
+                    'basis will be set to n_basis= %d' % n_basis)
+    elif isinstance(self.n_basis, int):
+      n_basis = self.n_basis
+    else:
+      raise ValueError("n_basis should be an integer, instead it is of type %s"
+                       % type(self.n_basis))
+
+    if n_basis > n_triplets:
+      n_basis = n_triplets
+      warnings.warn("The selected number of basis is greater than the number "
+                    "of points, only n_basis = %d will be generated" %
+                    n_basis)
+
+    basis = np.zeros((n_basis, n_features))
+
+    # get all positive and negative pairs with lowest index first
+    # np.array (2*n_triplets,2)
+    triplets_pairs_sorted = np.sort(np.vstack((triplets[:, [0, 1]],
+                                               triplets[:, [0, 2]])),
+                                    kind='stable')
+    # calculate all unique pairs and their indices
+    uniqPairs, indices = np.unique(triplets_pairs_sorted, return_inverse=True,
+                                   axis=0)
+    # calculate L2 distance according to bases only for unique pairs
+    diff = X[uniqPairs[:, 0], :] - X[uniqPairs[:, 1], :]
+
+    diff_pos = diff[indices[:n_triplets], :]
+    diff_neg = diff[indices[n_triplets:], :]
+
+    rng = check_random_state(self.random_state)
+
+    start = 0
+    finish = 0
+
+    while(finish != n_basis):
+
+      # select n_features positive differences
+      d_pos = diff_pos[rng.choice(n_triplets,
+                                  size=n_features, replace=False), :]
+
+      # select n_features negative differences
+      d_neg = diff_neg[rng.choice(n_triplets,
+                                  size=n_features, replace=False), :]
```

Comment:
A simpler alternative is to extract the positive and the negative pair from the same triplet (hence drawing only `n_triplets` rather than `2*n_triplets`). This probably works equally good?

Reply:
Actually that also makes more sense!